proposals/0015-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,391 @@
+`reward calculation phase`.
+
+On Solana Mainnet Beta with ~550K active stake accounts, the 90% cut-off of the
+epoch reward time for all the nodes is around 40 seconds. Note that this time
+includes both reward computation and reward distribution. The actual reward
+computation time could be less. However, to be conservative, we can safely
+assume the reward computation time to 40s.
+
+If we target 10% of the total block time for reward computation, this will give
+us `N=1000`, i.e. 40s/(400ms*10%). This parameter will be feature-gated and
+maybe updated in future.
+
+At the end of the `reward calculation phase`, aka. `Epoch_start + N` block
+height, the reward computation process is "barrier synchronized" with the bank
+replay process. If the reward calculation is slow and hasn't complete the
+computation at block-height `Epoch_start + N`, the bank replay process will
+block and wait until the reward computation is finished, before processing any
+blocks with height greater than `N`.
+
+After the full rewards result is calculated, the result is partitioned into
+distribution chunks, which will be distributed during the `reward distribution`
+phase.
+
+To ensure that each block distributes a subset of the rewards in a
+deterministic manner for the current epoch, while also randomizing the
+distribution across different epochs, the partitioning of all rewards will be
+done as follows.
+
+First, the reward results are sorted by account's Pubkey, and randomly shuffled
+with a fast `rng` seeded by current epoch. The shuffled reward results are then
+divided into `M` chunks. This process will ensure that the reward distribution
+is deterministic for the current epoch, while also introducing a degree of
+randomness between epochs.
+
+To minimize the impact on block processing time during the reward distribution
+phase, a total of 4,096 accounts will be distributed per block. The total
+number of blocks needed to distributed rewards is given by the following
+formula to avoid using floating point value arithmetic:
+
+```
+M = ((4096 - 1)+num_stake_accounts)/4096
+```
+
+### `EpochRewardHistory` Sysvar Account
+
+`EpochRewardHistory` sysvar account maintains a history of epoch rewards. The
+internal format of `EpochRewardHistory` is an array of `RewardHistoryEntry`,
+where each entry corresponds to the reward for one particular epoch. The
+maximum depth of the history is chosen to be the same as the depth of stake
+history for consistency, which is 512.
+
+The layout of `EpochRewardHistory` sysvar is shown in the following pseudo code.
+
+```
+struct RewardHistoryEntry {
+   total_reward_in_lamport: u64,          // total rewards for this epoch
+   distributed_reward_in_lamport: u64,    // already distributed reward amount
+   num_stake_accounts: u64,  // number of reward receiving stake accounts
+   total_vote_points: u128,  // accrued vote points in the epoch
+   root_hash: Option<Hash>,  // hash computed from all EpochRewardReserves
+}
```

Comment:
if we want to do the the hash-to-bucket thing, i think we're going to want to retain some parameters of the reward period here as well.  atm i'm thinking at least these

```rust
    num_reward_buckets: u64, /* can probably be more like `u16` or `u32` in reality */
    epoch_seed: u64, /* seed for hash function used to hash account addresses into buckets */
    /* maybe */
    hash_function_id: u8, /* in case we need to change it for some reason? */
```